keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,240 @@
+---
+title: ExecutionHook 
+authors:
+  - "@jingxu97"
+  - "@xing-yang"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+  - sig-node
+  - sig-architecture
+reviewers:
+  - "@saad-ali"
+  - "@thockin"
+approvers:
+  - "@thockin"
+  - "@saad-ali"
+editor: TBD
+creation-date: 2019-1-20
+last-updated: 2019-1-20
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+ExecutionHook API change
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+Volume snapshot support was introduced in Kubernetes v1.12 as an alpha feature. 
+In the alpha implementation of snapshots for Kubernetes, there is no snapshot consistency
+guarantees beyond any guarantees provided by storage system (e.g. crash consistency).
+
+This proposal is aimed to address that limitation by providing an `ExecutionHook`
+in the `Container` struct. The snapshot controller will look up this hook before
+taking a snapshot and execute it accordingly.
+
+## Motivation
+
+The volume snapshot feature allows creating/deleting volume snapshots, and the
+ability to create new volumes from a snapshot natively using the Kubernetes API.
+
+However, application consistency is not guaranteed. An user has to figure out how
+to quiece an application before taking a snapshot and unquiece it after taking
+the snapshot.
+
+So we want to introduce an `ExecutionHook` to facilitate the quiesce and unquiesce
+actions when taking a snapshot. There is an existing lifecycle hook in the `Container`
+struct. The lifecycle hook is called immediately after a container is created or
+immediately before a container is terminated. The proposed execution hook is not
+tied to the start or termination time of the container. Instead it is called before
+or after taking a snapshot while the container is still running.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container.
+Specifically it can be used to perform a quiese (freeze) operation before a volume
+snapshot is taken and then perform an unquiesce (thaw) operation after a volume
+snapshot is taken to resume the application.
+
+## Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+## Proposal
+
+An `ExecutionHook` is defined in the following.
+
+```
+// ExecutionHook defines a specific action that should be taken with timeout
+type ExecutionHook struct {
```

Comment:
Did you consider adding new entries to the already existing "lifecycle" section of the container spec?

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#lifecycle-v1-core

They do extremely similar things.

Reply:
That's a good point. We have not considered this.  Will look into it.

Reply:
Discussed with Jing about this. We added Alternatives which would include the new ExecutionHook directly inside "Lifecycle".  Please take a look.  Thanks.